File: algebra/blazingdb-calcite-application/src/main/java/com/blazingdb/calcite/application/Chrono/Chronometer.java
package com.blazingdb.calcite.application.Chrono;

import static java.util.concurrent.TimeUnit.NANOSECONDS;

import java.util.concurrent.TimeUnit;

public class Chronometer {
	private final Watch watch;
	private boolean isRunning;
	private long startTime;
	private long elapsedTime;

	public static Chronometer
	makeUnstarted() {
		return new Chronometer();
	}

	public static Chronometer
	makeStarted() {
		return makeUnstarted().start();
	}

	Chronometer() { this.watch = Watch.internalWatch(); }

	Chronometer(Watch watch) { this.watch = watch; }

	public boolean
	isRunning() {
		return isRunning;
	}

	public Chronometer
	start() {
		Check.state(!isRunning);
		isRunning = true;
		startTime = watch.read();
		return this;
	}

	public Chronometer
	stop() {
		long stopTime = watch.read();
		Check.state(isRunning);
		isRunning = false;
		elapsedTime += stopTime - startTime;
		return this;
	}

	public long
	elapsed() {
		return isRunning ? watch.read() - startTime + elapsedTime : elapsedTime;
	}

	public long
	elapsed(TimeUnit timeUnit) {
		return timeUnit.convert(elapsed(), NANOSECONDS);
	}

	public Chronometer
	reset() {
		isRunning = false;
		elapsedTime = 0;
		return this;
	}
}

File: tests/BlazingSQLTest/DemoTest/chronometer.py
import time


class Chronometer:
    @staticmethod
    def makeUnstarted():
        return Chronometer()

    @staticmethod
    def makeStarted():
        return Chronometer.makeUnstarted().start()

    @staticmethod
    def show(chronometer, message):
        elapsed = chronometer.elapsed()
        Chronometer.items.append((message, elapsed))
        print("\033[32mTime: %s: %dms\033[0m" % (message, elapsed))

    items = []

    @staticmethod
    def show_resume():
        print("\033[32mResume")
        for item in Chronometer.items:
            print("%s: %dms" % item)
        print("\033[0m")

    def __init__(self):
        self.watch = Watch()
        self.startTime = 0
        self.elapsedTime = 0

    def start(self):
        self.startTime = self.watch.read()
        return self

    def stop(self):
        stopTime = self.watch.read()
        self.elapsedTime += stopTime - self.startTime
        return self

    def elapsed(self):
        return self.watch.read() - self.startTime + self.elapsedTime

    def reset(self):
        self.elapsedTime = 0
        return self


class Watch:
    def read(self):
        return int(time.time() * 1000)
